assets/prototype/domain/eventEditor/containers/mutations/useEntityMutation.js

```diff
@@ -0,0 +1,146 @@
+import { useApolloClient } from '@apollo/react-hooks';
+import { ApolloError } from 'apollo-client';
+import { useState } from '@wordpress/element';
+
+import { mutations } from './';
+import useMutators from './useMutators';
+
+/**
+ * @param {string} type Entity type name
+ */
+const useEntityMutation = (type, id = '') => {
+	const client = useApolloClient();
+	const [result, setResult] = useState({ called: false, loading: false });
+	const mutators = useMutators();
+
+	/**
+	 * @param {string} mutation Type of mutation - CREATE|UPDATE|DELETE
+	 */
+	const getMutation = (mutationType = 'UPDATE') => {
+		// For example "CREATE_DATETIME"
+		const mutaton = `${mutationType}_${type.toUpperCase()}`;
+		return mutations[mutaton];
+	};
+
+	/**
+	 * @param {string} mutation Type of mutation
+	 * @param {object} data     Mutation data
+	 */
+	const getMutationOptions = (mutationType, data = {}) => {
+		// e.g. "datetmeMutator"
+		const key = `${type.toLowerCase()}Mutator`;
+		const { [key]: mutator } = mutators;
+		const { variables, optimisticResponse, update } = mutator(mutationType, data);
+
+		return { variables, optimisticResponse, update };
+	};
+
+	/**
+	 * @param {object} data
+	 */
+	const getCreateMutation = (data) => {
+		const mutationType = 'CREATE';
+		const mutation = getMutation(mutationType);
+		const options = getMutationOptions(mutationType, data);
+
+		return { mutation, ...options };
+	};
+
+	/**
+	 * @param {string} id Entity id
+	 */
+	const getUpdateMutation = (data) => {
+		const mutationType = 'UPDATE';
+		const mutation = getMutation(mutationType);
+		const options = getMutationOptions(mutationType, { ...data, id });
+
+		return { mutation, ...options };
+	};
+
+	/**
+	 *
+	 */
+	const getDeleteMutation = () => {
+		const mutationType = 'DELETE';
+		const mutation = getMutation(mutationType);
+		const options = getMutationOptions(mutationType, { id });
+
+		return { mutation, ...options };
+	};
```

Comment:
I love the readability here :1st_place_medal:

Reply:
ya me too